Core: Fix RSE `supported_checksums` attribute parsing; Fix #5155

akselaase opened the pull request:
The function `rse.py::parse_checksum_support_attribute` misbehaved when
being called with arguments of type `str` instead of the assumed
`list[str]`.

This commit adds explicit type hints to functions related to parsing the
`supported_checksums` attribute, and fixes errors at call sites passing
wrong types.

<!-- Please read https://rucio.cern.ch/documentation/contributing before submitting a pull request -->

joeldierkes commented on lib/rucio/core/rse.py:
```diff
@@ -749,14 +759,14 @@ def parse_checksum_support_attribute(checksum_attribute):
 
 @read_session
-def get_rse_is_checksum_supported(checksum_name, rse_id=None, session=None):
+def get_rse_is_checksum_supported(checksum_name: str, rse_id: str = None, session=None):
```
`rse_id` is `Optional[str]`

akselaase replied:
I actually think this should be a required argument (so remove the default `None` value). You have `is_checksum_valid` for the RSE-independent case.

joeldierkes commented on lib/rucio/core/rse.py:
```diff
@@ -721,22 +718,35 @@ def get_rse_attributes(rse_id, session=None):
-def get_rse_supported_checksums(rse_id, session=None):
+def get_rse_supported_checksums(rse_id: str, session=None):
     """
     Retrieve from the DB and parse the RSE attribute defining the checksum supported by the RSE
```
`:returns:`, `:params:` comments

joeldierkes commented on lib/rucio/core/rse.py:
```diff
@@ -721,22 +718,35 @@ def get_rse_attributes(rse_id, session=None):
     """
     Parse the RSE attribute defining the checksum supported by the RSE
     :param rse_attributes: attributes retrieved using list_rse_attributes
     """
```
`:returns:` missing

joeldierkes commented on lib/rucio/core/rse.py:
```diff
@@ -721,22 +718,35 @@ def get_rse_attributes(rse_id, session=None):
+    return parse_checksum_support_attribute(first_value)
+
+
+def get_rse_supported_checksums_from_attributes(rse_attributes: Dict[str, Any]):
```
Return type annotation missing

akselaase replied:
I haven't added them since they're inferred from the return of `parse_checksum_support_attribute`, would you prefer them being explicitly added everywhere?

joeldierkes replied:
For the moment I'd suggest to rather be descriptive, so yes. We can talk about it later.

joeldierkes commented on lib/rucio/core/rse.py:
```diff
@@ -749,14 +759,14 @@ def parse_checksum_support_attribute(checksum_attribute):
+        supported_checksum_list = checksum_attribute.split(',')
         if 'none' in supported_checksum_list:
             return []
         return supported_checksum_list
```
nit: This `else` indentation is unnecessary, since there is an early return.

joeldierkes commented on lib/rucio/tests/test_rse.py:
```diff
@@ -218,6 +221,65 @@ def test_empty_rse(self):
+        assert parse_checksum_support_attribute('md5') == ['md5']
+        assert parse_checksum_support_attribute('md5,adler32') == ['md5', 'adler32']
+
+    @pytest.mark.dirty
```
How about a `del_rse(rse_id)` instead of the dirty marker?

joeldierkes commented on lib/rucio/core/rse.py:
```diff
@@ -19,14 +19,14 @@
 from io import StringIO
 from re import match
-from typing import TYPE_CHECKING
+from typing import Any, List, Dict, Optional, Sequence
```
Alphabetical sorting?

joeldierkes reviewed (dismissed):
This is a good Draft PR. :wink:

akselaase commented:
Currently waiting on #5673 and maybe a PR from @joeldierkes to fix caching behaviour before updating this.

rcarpa commented on lib/rucio/core/rse.py:
```diff
@@ -782,13 +795,32 @@ def parse_checksum_support_attribute(checksum_attribute):
 
 
+@read_session
+def get_rse_is_checksum_supported(checksum_name: str, rse_id: str = None, session=None) -> bool:
```
this function isn't used anywhere. Why add it ? Do you intend to use it later ?

akselaase replied:
Leftovers from before the rebase, removed.

akselaase committed:
Core: Fix RSE `supported_checksums` attribute parsing; Fix #5155

The function `rse.py::parse_checksum_support_attribute` misbehaved when
being called with arguments of type `str` instead of the assumed
`list[str]`.

This commit adds explicit type hints to prevent call-site errors, and
fixes the implementation of `parse_checksum_support_attribute`.

rcarpa commented on lib/rucio/core/rse.py:
```diff
@@ -782,10 +783,12 @@ def parse_checksum_support_attribute(checksum_attribute):
     if not checksum_attribute:
         return GLOBALLY_SUPPORTED_CHECKSUMS
+
+    supported_checksum_list = checksum_attribute.split(',')
```
Why are you getting rid of strip() ?
